### PR TITLE
PipeNetDeviceComponent refactor

### PR DESCRIPTION
**Pull request by collinlunn**
Pumps, vents, and such no longer inherit from PipeNetDeviceComponent, they are their own components.

Should make it easier to add UI, since you can put in a shared component for pumps/etc. 

Also other code quality improvements.

### CODE REVIEW

File: Content.Server/GameObjects/Components/Atmos/Piping/PipeNetDeviceComponent.cs [hunk 1]
```diff
@@ -27,6 +29,12 @@ public override void OnRemove()
             LeaveGridAtmos();
         }
 
+        public void Update()
+        {
+            var message = new PipeNetUpdateMessage();
+            SendMessage(message);
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Atmos/Piping/PipeNetDeviceComponent.cs:35]: Creating the same empty message per pipe net device every update tick is expensive. Try caching the empty message and sending it over and over as needed, or use interfaces to update the rest of components instead.

File: Content.Server/GameObjects/Components/Atmos/Piping/Pumps/BasePumpComponent.cs [hunk 2]
```diff
@@ -33,87 +34,88 @@ public bool PumpEnabled
...
-                Logger.Error($"{typeof(BasePumpComponent)} on entity {Owner.Uid} did not have a {nameof(NodeContainerComponent)}.");
+                Logger.Error($"{typeof(BasePumpComponent)} on {Owner?.Prototype?.ID}, Uid {Owner?.Uid} did not have a {nameof(NodeContainerComponent)}.");
                 return;
             }
             var pipeNodes = container.Nodes.OfType<PipeNode>();
-            _inletPipe = pipeNodes.Where(pipe => pipe.PipeDirection == _inletDirection).FirstOrDefault();
-            _outletPipe = pipeNodes.Where(pipe => pipe.PipeDirection == _outletDirection).FirstOrDefault();
+            _inletPipe = pipeNodes.Where(pipe => pipe.PipeDirection == _initialInletDirection).FirstOrDefault();
+            _outletPipe = pipeNodes.Where(pipe => pipe.PipeDirection == _initialOutletDirection).FirstOrDefault();
             if (_inletPipe == null | _outletPipe == null)
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Atmos/Piping/Pumps/BasePumpComponent.cs:116]:
```suggestion
            if (_inletPipe == null || _outletPipe == null)
```

**Review by Zumorica — CHANGES_REQUESTED** [review]
Instead of using `Owner.EnsureComponent`, try using `Owner.EnsureComponentWarn`! That way it'll print a warning if the entity *was* missing the component in question, which is honestly preferred to silently fixing it.

File: Content.Server/GameObjects/Components/Atmos/Piping/GasCanisterPortComponent.cs [hunk 3]
```diff
@@ -1,54 +1,43 @@
...
-                    .FirstOrDefault();
-                if (anchoredCanister != null)
+                var entities = snapGrid.GetLocal();
+                foreach (var entity in entities)
                 {
-                    anchoredCanister.TryConnectToPort();
+                    if (entity.TryGetComponent<GasCanisterComponent>(out var canister) && canister.Anchored && !canister.ConnectedToPort)
+                    {
+                        canister.TryConnectToPort();
+                    }
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Atmos/Piping/GasCanisterPortComponent.cs:40]: Should return if connected if you want to replicate the old behavior.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Review by metalgearsloth — APPROVED** [review]

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**